src/main/java/org/jahia/modules/appshell/Main.java

```diff
@@ -144,49 +249,145 @@ private void setCustomAttributes(JahiaUser currentUser, HttpServletRequestWrappe
+    interface ValueProcessor {
+        void processValues(List<String> values, AppInfo appInfo);
     }
 
-    private List<String> getBundleScripts(Bundle bundle, String appName, String key) throws IOException, JSONException {
-        JSONObject pkgJson = new JSONObject(IOUtils.toString(bundle.getResource(PACKAGE_JSON) != null ? bundle.getResource(PACKAGE_JSON) : bundle.getResource(JAHIA_JSON)));
-        boolean hasExtend = pkgJson.has(JAHIA) && pkgJson.getJSONObject(JAHIA).has(key) && pkgJson.getJSONObject(JAHIA).getJSONObject(key).has(appName);
-        if (hasExtend) {
-            Object res = pkgJson.getJSONObject(JAHIA).getJSONObject(key).get(appName);
-            if (res instanceof JSONArray) {
-                JSONArray jsonArray = (JSONArray) res;
-                List<String> l = new ArrayList<>();
-                int s = jsonArray.length();
-                for (int i = 0; i < s; i++) {
-                    l.add(jsonArray.getString(i));
-                }
-                return l;
-            } else {
-                return Collections.singletonList(res.toString());
+    private void processUrlPatterns(Map<String, AppInfo> newAppInfos, JSONObject pkgJson) throws JSONException {
+        JSONObject appsUrlPatterns = pkgJson.getJSONObject(JAHIA).getJSONObject(URLPATTERNS);
+        processAppValues(newAppInfos, appsUrlPatterns, (values, appInfo) -> {
+            for (String urlPattern : values) {
+                Pattern compiledUrlPattern = Pattern.compile(urlPattern);
+                appInfo.getCompiledUrlPatterns().add(compiledUrlPattern);
+            }
+        });
+    }
+
+    private void processKeyScripts(Map<String, AppInfo> newAppInfos, JSONObject pkgJson, Bundle bundle, String key) throws JSONException {
+        JSONObject scriptObj = pkgJson.getJSONObject(JAHIA).getJSONObject(key);
+        processAppValues(newAppInfos, scriptObj, (values, appInfo) -> {
+            for (String script : values) {
+                validateScript(bundle, appInfo, script, key);
+            }
+        });
+    }
+
+    private void validateScript(Bundle bundle, AppInfo appInfo, String script, String key) {
+        if(bundle.getResource(script) == null) {
+            logger.error("Application {} declared {} has entry point but file is not found, skipping it from {}", bundle.getSymbolicName(), script, key);
+            return;
+        }
+        List<String> scripts = appInfo.getScripts().get(key);
+        if (scripts == null) {
+            scripts = new ArrayList<>();
+        }
+        if (script.startsWith("/")) {
+            scripts.add(script);
+        } else {
+            scripts.add("/modules/" + bundle.getSymbolicName() + "/" + script);
+        }
+        appInfo.getScripts().put(key, scripts);
+    }
```

Comment:
I would move this code directly inside the callback:
```
processAppValues(newAppInfos, scriptObj, (values, appInfo) -> {
            for (String script : values) {
                validateScript(bundle, appInfo, script, key);
            }
        });
```
I would find it more readable.

Reply:
Good suggestion, I'll do that.